Automattic/wp-calypso
# Checkout: Fix shopping cart plan overlap notices

When a site has a Jetpack Personal plan and a user attempted to purchase Jetpack Scan, the checkout `PrePurchaseNotices` component was incorrectly displaying that Jetpack Scan is already included in their Jetpack Personal plan, which Jetpack Scan is not included in the Jetpack Personal plan (only Backup and Anti-Spam). (See image)

<img width="489" alt="Screenshot 2023-06-12 at 14 56 27" src="https://github.com/Automattic/jpop-issues/assets/29981758/d63ec3ec-67d8-4c37-9359-34629b2306bf">

I then discovered that various other plan/product overlaps were missing from the notice, such as for plans that include Search, Boost, Social, and VideoPress.



Closes: https://github.com/Automattic/jpop-issues/issues/8345

Asana task: 1198218726984184-as-1204805969978494/f

## Proposed Changes

- This PR fixes the issue with the Jetpack Personal plan and Jetpack Scan (also described in issue https://github.com/Automattic/jpop-issues/issues/8345) 
- Also I refactored the `PrePurchaseNotices` component to now check for plan overlaps with Search, Boost, Social, and VideoPress products in the cart.
- Also I converted the `PrePurchaseNotices` component file to TypeScript (.tsx).

## Testing Instructions

- First apply patch D120104-code to your sandbox, and sandbox `public-api.wordpress.com`.
- In Store Admin, add Jetpack Personal to your connected Jetpack site.
- Checkout this PR, ( `git fetch && git checkout fix/checkout-pre-purchase-notice && yarn start` )
- OR: Use the Calypso Live (direct link) below.
- Go to `/checkout/:SITE/jetpack_scan`
-  Verify you no longer get a notice that Jetpack Scan is already included in the Jetpack Personal plan (see image above and issue https://github.com/Automattic/jpop-issues/issues/8345) .
- Go back to Store Admin and remove the Jetpack Personal subscription and replace it with Jetpack Complete.
- Now test the following products in the cart, one by one, and verify you **see the notice** stating that:
"**{Product_name} is included in your Jetpack Complete subscription, so it was removed from the cart.**"
    - `/checkout/:SITE/jetpack_boost_yearly`
    - `/checkout/:SITE/jetpack_social_basic_yearly`
    - `/checkout/:SITE/jetpack_videopress`
- You can also check:
    - `/checkout/:SITE/jetpack_anti_spam`
    - `/checkout/:SITE/jetpack_scan_monthly`
    - `/checkout/:SITE/jetpack_backup_daily` (You'll get a pre-purchase warning, but it will not be removed from the cart.)
    - `/checkout/:SITE/jetpack_search` (You'll get a pre-purchase warning, but it will not be removed from the cart.)
    - `/checkout/:SITE/jetpack_backup_t1_yearly` (You'll get a different notice that adds a Backup Add-on product instead.)
    - 


## Pre-merge Checklist
